lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+    }
+
+    thetas
+}
+
+/// Number of fast clock counts between two consecutive
+/// TimeStamps. This requires that `tstamps[0]` is more recent than
+/// `tstamps[1]` but otherwise imposes no restrictions on them. For
+/// instance, they can be from different processing periods and this
+/// will still count the number of counts between them, accounting for
+/// overflow wrapping. Both TimeStamps must be valid for this to be
+/// used.
+///
+/// # Arguments
+///
+/// * `tstamps_mem` - Recorded TimeStamps.
+/// * `overflow_count` - Max timestamp value.
+fn tstamps_diff(tstamps_mem: &[TimeStamp; 2], overflow_count: u32) -> u32 {
+    debug_assert!(tstamps_valid_count(tstamps_mem) == 2);
+
+    if tstamps_mem[0].sequences_old == tstamps_mem[1].sequences_old {
+        return tstamps_mem[0].count - tstamps_mem[1].count;
+    }
+
+    let rem0: u32 = tstamps_mem[0].count;
+    let rem1: u32 = overflow_count - tstamps_mem[1].count;
+    let empty_sequences =
+        tstamps_mem[1].sequences_old - tstamps_mem[0].sequences_old - 1;
+
+    rem0 + rem1 + overflow_count * empty_sequences as u32
+}
+
+/// Increment `sequences_old` in each TimeStamp of `tstamps_mem`.
+fn increment_tstamp_sequence(tstamps_mem: &mut [TimeStamp; 2]) {
+    for tstamp in tstamps_mem {
+        tstamp.new_sequence();
+    }
+}
+
+/// Compute the phase (in radians) for a phase given in counts
+/// relative to some period in counts.
+///
+/// # Arguments
+///
+/// * `theta_count` - Phase in counts. This can be greater than the
+/// period in counts.
+/// * `fscale` - Frequency scaling factor for harmonic demodulation.
+/// * `period_count` - Number of counts in 1 period.
+/// * `phase_count` - Phase offset in counts.
+///
+/// # Latency: 138ns
+fn real_phase(
+    theta_count: u32,
+    fscale: u32,
+    period_count: u32,
+    phase_count: u32,
+) -> f32 {
+    let total_angle =
+        modulo::<u32>(theta_count * fscale + phase_count, period_count);
+    2. * PI * (total_angle as f32 / period_count as f32)
+}
+
+/// Arithmetic modulo operation. This requires that the dividend is
+/// non-negative and the divisor is positive. Although this may work
+/// in some other cases, these cases are not supported.
+fn modulo<T: SubAssign + Ord + Copy + From<u8>>(dividend: T, divisor: T) -> T {
+    debug_assert!(dividend >= T::from(0u8) && divisor > T::from(0u8));
+
+    let mut rem: T = dividend;
+    while rem > divisor {
+        rem -= divisor;
+    }
+    rem
+}
+
+/// Filter in-phase and quadrature signals with the IIR biquad filter.
+///
+/// TODO this currently does not offer enough filtering flexibility. For
+/// instance, we might want to filter with two consecutive biquads.
+///
+/// # Arguments
+///
+/// `i` - In-phase signals.
+/// `q` - Quadrature signals.
+/// `iir` - IIR filters for the in phase (index 0) and quadrature
+/// (index 1) signals.
+/// `iirstate` - State of each IIR filter.
+///
+/// Latency (ADC batch size = 16): 3.5us
+fn filter(
```

Comment:
I agree - I think we could actually just do this in-place with a `map()`. It would be cleaner and likely more performant. Just mutably borrow the original array and modify it in-place.